src/txn_builder.rs

```diff
@@ -1117,92 +1137,128 @@ impl<L: Ledger> TransactionState<L> {
             .fold(0, |sum, (_, amount)| sum + *amount);
 
         // find input records of the asset type to spend (this does not include the fee input)
-        let (input_records, change) = self.find_records(
+        let records = self.find_records(
             spec.asset,
-            &spec.owner_keypair.pub_key(),
+            spec.owner_key_pairs,
             FreezeFlag::Unfrozen,
             total_output_amount,
             None,
         )?;
-        let asset = input_records[0].0.asset_def.clone();
 
-        // prepare inputs
-        let mut inputs = vec![];
-        for (ro, uid) in input_records.iter() {
-            let witness = self.get_merkle_proof(*uid);
-            inputs.push(TransferNoteInput {
-                ro: ro.clone(),
-                acc_member_witness: witness,
-                owner_keypair: spec.owner_keypair,
-                cred: None, // TODO support credentials
-            })
+        let asset = records[0].1[0].0.asset_def.clone();
+
+        let mut inputs = Vec::new();
+        let mut input_ros = Vec::new();
+        let mut fee_input = None;
+        let mut change_ro = None;
+        for (owner_key_pair, input_records, change) in &records {
+            // prepare inputs
+            for (ro, uid) in input_records.iter() {
+                let witness = self.get_merkle_proof(*uid);
+                inputs.push(TransferNoteInput {
+                    ro: ro.clone(),
+                    acc_member_witness: witness,
+                    owner_keypair: owner_key_pair,
+                    cred: None, // TODO support credentials
+                });
+                input_ros.push(ro.clone());
+            }
+            if fee_input.is_none() {
+                if let Ok(input) = self.find_fee_input(owner_key_pair, spec.fee) {
+                    fee_input = Some(input);
+                }
+            }
+
+            // change in the asset type being transfered (not fee change)
+            if *change > 0 {
+                let me = owner_key_pair.pub_key();
+                change_ro = Some(RecordOpening::new(
+                    rng,
+                    *change,
+                    asset.clone(),
+                    me,
+                    FreezeFlag::Unfrozen,
+                ));
+            }
         }
-        let fee_input = self.find_fee_input(spec.owner_keypair, spec.fee)?;
 
         // prepare outputs, excluding fee change (which will be automatically generated)
-        let mut outputs = vec![];
+        let mut outputs = Vec::new();
         for (pub_key, amount) in spec.receivers {
             outputs.push(RecordOpening::new(
-                rng,
+                &mut rng.clone(),
```

Comment:
Why is this needed?

Reply:
Oh it was needed previously, but not anymore. I've reverted it.